solutions/usaco-282.mdx

```diff
@@ -0,0 +1,148 @@
+---
+id: usaco-282
+source: USACO Silver 2013 Open
+title: What's Up With Gravity
+author: Chuyang Wang
+---
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_gravity.html)
+
+
+## Explanation
+
+Since this problem can be reduced to a shortest path problem with edge lengths
+of 0 or 1, we can use BFS. A transition edge has the weight 1 if the gravity is
+flipped or 0 if Captain Bovidian only moves left or right. We firstly add all
+reachable blocks from left and right to the queue. Each element in the queue is
+a 4-tuple containing the row and column of the block, the number of flips
+needed to reach it, and the direction of gravity. Then, for each block in the
+queue, we change the gravity and perform another search for the left and right,
+eventually adding new blocks to the end of the queue. This way, we will always
+check blocks that require fewer flips before moving forward.
+
+During the search process, we store the number of flips needed to get to a
+certain block. We only visit blocks that are not visited yet.
+
+**Time Complexity:** $\mathcal{O}(N^2)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+/**
+ * With the given direction of gravity, calculate the row on which captain
+ * will fall on.
+ * @param direc Direction of the gravity, either 1 or -1
+ */
+int fall_to_row(int r, int c, int direc, const vector<vector<char>> &grid) {
+	while (grid[r][c] == '.' || grid[r][c] == 'C' || grid[r][c] == 'D') {
+		if (grid[r][c] == 'D') {
+			return r;
+		}
+		if (grid[r + direc][c] == '#') {
+			return r;
+		}
+		r += direc;
+	}
+	return -1;
+}
+
+/**
+ * Find all reachable blocks from [r, c] by going left and right, without
+ * flipping the gravity.
+ * @param flips Number of flips already performed to get to [r, c]
+ * @param direc Current direction of the gravity, either 1 or -1
+ */
+void search_horizontally(
+	int r, int c, int flips, int direc,
+	queue<array<int, 4>> &bfs,
+	vector<vector<int>> &flip_count,
+	const vector<vector<char>> &grid
+) {
+	if (r <= 0 || grid[r][c] == '-' || flip_count[r][c] >= 0) {
```

Comment:
maybe a comment to the effect of "if OOB, [whatever checking for a hypen does], or we've alr visited, don't go any further"

Reply:
> whatever checking for a hypen does

It just checks out of bound

`// border is marked as '-'`